### ISSUE TITLE
Issue when using bulk api to insert data

### ISSUE THREAD

**Issue by AAraKKe** [simple-salesforce/simple-salesforce#366]
Hi! Thanks a lot for this great tool! 
I just discovered it and I am doing some tests to try and use it for bulk APIs updates. I am sorry if this has been posted or clarified before, I could not find anything. 

I can do mostly everything, bulk querying included, but when I try to make a bulk insert the call fails with the error

`{'exceptionCode': 'InvalidBatch', 'exceptionMessage': 'Records not processed'}`

The batch I am running as a test is like this:

```python
bulk = []
for dd in range(10):
    tldate = dt.datetime(2019,1,1) + dt.timedelta(days=dd)
    teledata = {
        'DAU__c': 10,
        'MAU__c': 30,
        'WAU__c': 20,
        'AccountId__c': '0010o00002haQFOAA2',
        'Telemetry_Date__c': tldate.isoformat()
    }
    bulk.append(teledata)

client.bulk.TelemetryData__c.insert(teledata)
```

Basically, I am trying to insert custom objects to a given account. If I create one by one it works fine but when using the bulk api to insert it the error appears. I am not sure what I am doing wrong. The ORG has the object active for bulk api usage. I can query it using bulk:

```python
client.bulk.TelemetryData__c.query('SELECT WAU__c, Name FROM TelemetryData__c LIMIT 10')
>>> [{'attributes': {'type': 'TelemetryData__c',
   'url': '/services/data/v38.0/sobjects/TelemetryData__c/a000o00004cn2Y4AAI'},
  'WAU__c': 20.0,
  'Name': 'TD-00000022'},
 {'attributes': {'type': 'TelemetryData__c',
   'url': '/services/data/v38.0/sobjects/TelemetryData__c/a000o00004cn2k0AAA'},
  'WAU__c': 20.0,
  'Name': 'TD-00000023'},
 {'attributes': {'type': 'TelemetryData__c',
   'url': '/services/data/v38.0/sobjects/TelemetryData__c/a000o00004cn2kAAAQ'},
  'WAU__c': 20.0,
  'Name': 'TD-00000024'}]
```

Those are the entires I just created one by one with no problem at all. Do you know what the issue might be? Thanks!

**Comment by jon-wobken** [comment]
@AAraKKe - If you navigate to Setup -> Bulk Data Load Jobs and tap the "View Request" and "View Result", can you paste what the request data and the result data that is being posted is and are you sure that the 'AccountId__c': '0010o00002haQFOAA2' is a valid field/account?

When you do it one by one, are you manually entering the data in the list or setting the loop to just run 1 loop? 

I'd also recommend changing your variable name from bulk just as a precaution. 

bulk_insert_records = []
for dd in range(10):
    tldate = dt.datetime(2019,1,1) + dt.timedelta(days=dd)
    teledata = {
        'DAU__c': 10,
        'MAU__c': 30,
        'WAU__c': 20,
        'AccountId__c': '0010o00002haQFOAA2',
        'Telemetry_Date__c': tldate.isoformat()
    }
    bulk_insert_records.append(teledata)

client.bulk.TelemetryData__c.insert(teledata)

**Comment by AAraKKe** [comment]
Hi @jon-wobken , thanks for the quick response. A couple of comments, if you notice in my message above, thee was a mistake in which I was inserting teledata, which is the dictionary, not the array for bulk. I just noticed that a bit after, sorry for that. Nonetheless, using the proper array still shows an error:

```
'InvalidBatch : Invalid batch format'
```

The account exists, yes. And I can easily create single entires without bulk using the same fields. About the output you request here is what it shows:

View Result
```
{"exceptionCode":"InvalidBatch","exceptionMessage":"Records not processed"}
```

View Request
```
[{"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}, {"DAU__c": 10, "MAU__c": 30, "WAU__c": 20, "AccountId__c": "0010o00002haQFOAA2", "Telemetry_Date__c": "2019-01-11T00:00:00"}]
```

PS: I noticed this does not use bulk api V2, isn't it? Is it possible to use it and I might have just missed it?

**Comment by jon-wobken** [comment]
@AAraKKe - Bulk API v2 is not currently supported in simple-salesforce. Do you have a restriction on the Telemetry_Date__c ? Looking at your data, it looks like the telemetry_date__c field is always the same. My best guess is either some validation rule/field restriction/backend process that is blocking the batch, potentially on that telemetry_date__c field since it looks like it never changes in your request data.

I just confirmed in one of my orgs that I don't have an issue with the bulk insert and inserting time data.

**Comment by jon-wobken** [comment]
@AAraKKe - Just wanted to follow up and see if you had any more detailed error information or if you were able to resolve this.

**Comment by MikeS-nth** [comment]
@AAraKKe  It's possible the issue is with how your date value is formatted in your code.  

Judging by:
`for dd in range(10):
    tldate = dt.datetime(2019,1,1) + dt.timedelta(days=dd)`
It looks like you're taking a start date and iterating it up by one day.  That tells me you don't care about time, yes?  Just date?

 I was running into this same issue, and solved it by converting my dates to strings.  Try adding:
`tldate = tldate.strftime('%Y-%m-%d')` before adding it to `teledata`

**Comment by AAraKKe** [comment]
Hi and sorry for the delay. When I use the date the parameter is sent as the isoformat of the date and it does not include the time as far as I have seen. I stopped researching this because we wanted to use version 2.0 of the bulk api. I ended up building my own wrapper for the needs we have. If this is working for everyone else maybe I was running into an issue myself, but doing it myself with the bulk api 2.0 worked just fine.

Sorry I could not be of more help to rack it down and thanks a lot for your help.
